### ISSUE TITLE
Guardness checker and recursive call on x0 in return-type of match x as x0

### ISSUE THREAD

**Issue by fakusb** [coq/coq#9761]
It seems that the well-guardedness checker does not know that in the return type of `match x as x0`, any subterm of `x0` is also a subterm of `x`?

I'm not sure if this is a bug or a feature request (and who even knows if this would be sound...)

The minimal example I came up with is (in coq 8.8 and master from less than three weeks ago):
```
Check (fix f x : {n & n >= 0} :=
             existT (fun n => n >= 0)
                    match x with
                      (S x1) => projT1 (f x1)
                    | _ => 0
                    end
                  match x
                  with
                    (S x1) => projT2 (f x1)
                  | _ => le_n 0
                  end).
```
produces the error
```
Error:
Recursive definition of f is ill-formed.
In environment
f : nat -> {n : nat & n >= 0}
x : nat
x0 : nat
x1 : nat
Recursive call to f has principal argument equal to "x1" instead of a subterm of "x".
Recursive definition is:
"fun x : nat =>
 existT (fun n : nat => n >= 0) match x with
                                | 0 => 0
                                | S x1 => projT1 (f x1)
                                end
   match x as x0 return (match x0 with
                         | 0 => 0
                         | S x1 => projT1 (f x1)
                         end >= 0) with
   | 0 => le_n 0
   | S x1 => projT2 (f x1)
   end".
```

(btw: changing the return clause to `match x` instead `x0` (and removing the bottommost cases with `_`) satisfies the guardedness checker.)

**Comment by gares** [comment]
Very tricky indeed. Let me just ping @barras @maximedenes

**Comment by fakusb** [comment]
I think this is the important part of the termination checker:
https://github.com/coq/coq/blob/27ab0d69e87fa3ce414cdaa205f7f6266693e263/kernel/inductive.ml#L930
Maybe separating of `check_rec_call` on `p` would be a solution? Then all the indices and `c_0` should be supplied as the stack instead of `[]`? But getting the indices by computing the type of `c_0` again seems a bit fishy...

**Comment by ppedrot** [comment]
I don't think this is sound in general. It intuitively looks like you can perform ill-founded recursive calls in dead return branches with this kind of criterion. (I don't have a proper example though.)

**Comment by fakusb** [comment]
@ppedrot Could you elaborate a bit more? To me, it is counter intuitive that the termination checker could accept an application of the return clause to the matched term 'in a vacuum' (in my example `(fun x0 => match x0 with .. end >= 0) x`), but that there would be a problem if this is used as a return clause in a match.

**Comment by JasonGross** [comment]
Duplicate of #9045

**Comment by ppedrot** [comment]
@JasonGross how is that a duplicate? In the linked issue, the problem stems from mutually recursive calls and the recursive argument does not appear in the return clause of any match. The current issue, contrastingly, proposes to extend the guard condition to return clauses, a feature for which I cannot even imagine a translation into recursors.

**Comment by JasonGross** [comment]
Ah, sorry, I did not read the other issue carefully enough.  I did an issue search for the guard checker and return clauses, this and the other issue were the only two to come up, and they looked superficially similar.  I shall aim to read more carefully in the future before declaring duplicates.

**Comment by JasonGross** [comment]
> a feature for which I cannot even imagine a translation into recursors.

It seems like you should be able to translate most uses into having the recursor compute a sigma of the return type and a thing of that type.  That is, any time you have fixpoint `f` with a body matching `context G[match x return @?P (f x) x with cn => bn end]` you should be able to replace it with the propositionally equal body `match x with cn => G[x => cn][bn] end` where `[x => cn]` means "replace x with cn"
